src/stream/mod.rs

```diff
@@ -471,3 +476,231 @@ impl Stream {
+/// [`AsyncWrite`].
+///
+/// Both `poll_read` and `poll_write` calls allocate temporary buffers, which results in an
+/// additional overhead.
+pub struct PollStream<'a> {
+    stream: Arc<Stream>,
+
+    read_fut: Option<Pin<Box<dyn Future<Output = Result<Vec<u8>>> + Send + 'a>>>,
+    write_fut: Option<Pin<Box<dyn Future<Output = Result<usize>> + Send + 'a>>>,
+    shutdown_fut: Option<Pin<Box<dyn Future<Output = Result<()>> + Send + 'a>>>,
+}
+
+impl PollStream<'_> {
+    /// Constructs a new `PollStream`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use webrtc_sctp::stream::{Stream, PollStream};
+    /// use std::sync::Arc;
+    ///
+    /// let stream = Arc::new(Stream::default());
+    /// let poll_stream = PollStream::new(stream);
+    /// ```
+    pub fn new(stream: Arc<Stream>) -> Self {
+        Self {
+            stream,
+            read_fut: None,
+            write_fut: None,
+            shutdown_fut: None,
+        }
+    }
+
+    /// Get back the inner stream.
+    pub fn into_inner(self) -> Arc<Stream> {
+        self.stream
+    }
+
+    /// Obtain a clone of the inner stream.
+    pub fn clone_inner(&self) -> Arc<Stream> {
+        self.stream.clone()
+    }
+
+    /// stream_identifier returns the Stream identifier associated to the stream.
+    pub fn stream_identifier(&self) -> u16 {
+        self.stream.stream_identifier
+    }
+
+    /// buffered_amount returns the number of bytes of data currently queued to be sent over this stream.
+    pub fn buffered_amount(&self) -> usize {
+        self.stream.buffered_amount.load(Ordering::SeqCst)
+    }
+
+    /// buffered_amount_low_threshold returns the number of bytes of buffered outgoing data that is
+    /// considered "low." Defaults to 0.
+    pub fn buffered_amount_low_threshold(&self) -> usize {
+        self.stream.buffered_amount_low.load(Ordering::SeqCst)
+    }
+
+    /// get_num_bytes_in_reassembly_queue returns the number of bytes of data currently queued to
+    /// be read (once chunk is complete).
+    pub(crate) async fn get_num_bytes_in_reassembly_queue(&self) -> usize {
+        // No lock is required as it reads the size with atomic load function.
+        let reassembly_queue = self.stream.reassembly_queue.lock().await;
+        reassembly_queue.get_num_bytes()
+    }
+}
+
+impl AsyncRead for PollStream<'_> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let mut temp_buf = vec![0; buf.remaining()];
```

Comment:
I would suggest using a buffer that has a constant size.
The user can, in theory, call `poll_read(cx, buffer_of_size_1024)`, which returns `Pending`, then call `poll_read(cx, buffer_of_size_2048)`.